### PR TITLE
Implement asymmetric sync for the .NET SDK

### LINKED ISSUE #2908: Implement Asymmetric Sync

### PR DESCRIPTION
**Pull request by LaPeste**
## Description
Implements asymmetric sync for the .NET SDK
Fixes #2908 

##  TODO

* [x] Changelog entry
* [x] Tests (if applicable)

### CODE REVIEW

File: Realm/Realm.Weaver/Extensions/TypeReferenceExtensions.cs [hunk 1]
```diff
@@ -46,29 +46,32 @@ SequencePoint GetPropSequencePoint()
         }
     }
 
-    public static bool IsRealmObjectDescendant(this TypeReference @this, ImportedReferences references)
+    private static bool IsTObjectDescendant(TypeReference target, params TypeReference[] instanceOf)
```
Comment (nirinchev) [inline Realm/Realm.Weaver/Extensions/TypeReferenceExtensions.cs:49]:
```suggestion
    private static bool IsDescendantOf(TypeReference @this, params TypeReference[] targetTypes)
```

I got a little confused as to what `target` and `instanceOf` mean. I'm not sure my suggestion is much better, but I think it's a step in the right direction.

File: Realm/Realm/DatabaseTypes/AsymmetricObject.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    /// The benefit of using <see cref="AsymmetricObject"/> is that the performance of each sync operation is much higher.
+    /// The drawback is that an <see cref="AsymmetricObject"/> is synced unidirectionally, so it cannot be queried.
+    /// You should use this base when you have a write-heavy use case.
+    /// If, instead you want to persist an object that you can also query against, use <see cref="RealmObject"/> instead.
+    /// RealmObjects and EmbeddedObjects can't link (or backlink) to AsymmetricObjects. AsymmetricObjects can only link to EmbeddedObjects.
+    /// </remarks>
+    /// <seealso cref="TODO andrea: add asymmetric sync link page here"/>
+    public class AsymmetricObject : RealmObjectBase, IAsymmetricObject
+    {
+        // TODO andrea: should this be only instantiated in a write transaction? Since we want to enforce them to be short-lived
```
Comment (nirinchev) [inline Realm/Realm/DatabaseTypes/AsymmetricObject.cs:34]: No - I think it's fine to instantiate them outside. An example is a sensor which records readings, but doesn't want to sync them one by one. Instead, the developer can accumulate readings from 10-30s in memory, then sync everything in a single transaction.

File: Realm/Realm/Dynamic/DynamicAsymmetricObject.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
+﻿////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
```
Comment (nirinchev) [inline Realm/Realm/Dynamic/DynamicAsymmetricObject.cs:3]:
```suggestion
// Copyright 2022 Realm Inc.
```

File: Realm/Realm/Dynamic/MetaRealmObject.cs [hunk 4]
```diff
@@ -29,6 +29,7 @@
 
 namespace Realms.Dynamic
 {
+    // TODO andrea: this class may need adjustment as there are no tests written yet
```
Comment (nirinchev) [inline Realm/Realm/Dynamic/MetaRealmObject.cs:32]: This should be tested in `Dynamic***Tests`.

File: Realm/Realm/Extensions/ReflectionExtensions.cs [hunk 5]
```diff
@@ -47,8 +48,26 @@ public static bool HasCustomAttribute<T>(this MemberInfo member)
...
+
         public static bool IsRealmObject(this Type type) => typeof(IRealmObject).IsAssignableFrom(type);
 
+        public static ObjectSchema.ObjectSchemaType GetRealmSchemaType(this Type type)
+        {
+            if (type.IsAsymmetricObject())
+            {
+                return ObjectSchema.ObjectSchemaType.TopLevelAsymmetric;
+            }
+            else if (type.IsEmbeddedObject())
```
Comment (nirinchev) [inline Realm/Realm/Extensions/ReflectionExtensions.cs:61]: You don't need the `else`-s here.

File: Realm/Realm/Native/SchemaObject.cs [hunk 6]
```diff
@@ -32,6 +32,6 @@ internal struct SchemaObject
         internal int properties_end;
 
         [MarshalAs(UnmanagedType.U1)]
```
Comment (nirinchev) [inline Realm/Realm/Native/SchemaObject.cs:34]: That should not be necessary anymore

File: Realm/Realm/Native/SchemaObject.cs [hunk 7]
```diff
@@ -32,6 +32,6 @@ internal struct SchemaObject
...
-        internal bool is_embedded;
+        internal byte table_type;
```
Comment (nirinchev) [inline Realm/Realm/Native/SchemaObject.cs:35]: You should be able to use the enum here, not the underlying type.

File: Realm/Realm/Realm.cs [hunk 8]
```diff
@@ -608,9 +621,11 @@ internal void ManageEmbedded(IEmbeddedObject obj, ObjectHandle handle)
             obj.SetManagedAccessor(new ManagedAccessor(this, handle, metadata), metadata.Helper, false, true);
         }
 
-        private void AddInternal(IRealmObject obj, Type objectType, bool update)
+        // TODO andrea: see if revert this to IRealmObject and have another AddInternal with different rules
+        private void AddInternal<T>(T obj, Type objectType, bool update)
+            where T : IRealmObjectBase
         {
-            if (!ShouldAddNewObject(obj))
+            if (!ShouldAddNewObject(obj) || obj.GetType().IsEmbeddedObject())
```
Comment (nirinchev) [inline Realm/Realm/Realm.cs:628]: If we call this with `EmbeddedObject`, this should be an error, shouldn't it?

File: Realm/Realm/Schema/ObjectSchema.cs [hunk 9]
```diff
@@ -35,6 +35,16 @@ namespace Realms.Schema
...
     public class ObjectSchema : IReadOnlyCollection<Property>
     {
+        // TODO andrea:
+        // 1 - move this to the right place
+        // 2- add documentation
+        public enum ObjectSchemaType : byte
+        {
+            TopLevel = 0,
+            Embedded = 1,
+            TopLevelAsymmetric = 2,
```
Comment (nirinchev) [inline Realm/Realm/Schema/ObjectSchema.cs:45]: If those are public, we should match the terminology we use for objects - i.e. `RealmObject`, `EmbeddedObject`, `AsymmetricObject`.

File: Realm/Realm/Schema/ObjectSchema.cs [hunk 10]
```diff
@@ -59,12 +69,22 @@ public class ObjectSchema : IReadOnlyCollection<Property>
...
         public bool IsEmbedded { get; }
 
-        private ObjectSchema(string name, bool isEmbedded, IDictionary<string, Property> properties)
+        // TODO andrea: review this documentation
+        /// <summary>
+        /// Gets a <see cref="ObjectSchemaType"/> indicating whether this <see cref="ObjectSchema"/> describes
+        /// a top level object, an embedded object or an asymmetric object.
+        /// </summary>
+        /// <value>The type of ObjectSchema.</value>
+        public ObjectSchemaType RealmSchemaType { get; }
```
Comment (nirinchev) [inline Realm/Realm/Schema/ObjectSchema.cs:81]: The name of this property is kind of weird. I'd suggest just `Type` or `ObjectType`.

File: Realm/Realm/Schema/ObjectSchema.cs [hunk 11]
```diff
@@ -35,6 +35,16 @@ namespace Realms.Schema
     [DebuggerDisplay("Name = {Name}, Properties = {Count}")]
     public class ObjectSchema : IReadOnlyCollection<Property>
     {
+        // TODO andrea:
+        // 1 - move this to the right place
+        // 2- add documentation
+        public enum ObjectSchemaType : byte
```
Comment (nirinchev) [inline Realm/Realm/Schema/ObjectSchema.cs:41]: The name of the enum is a little misleading as this is not the type of the objectschema but rather the type of the object the objectschema represents.

File: Realm/Realm/Schema/ObjectSchema.cs [hunk 12]
```diff
@@ -145,31 +165,54 @@ public class Builder : SchemaBuilderBase<Property>
...
+            /// </summary>
+            /// <param name="name">The name of the <see cref="ObjectSchema"/> this builder describes.</param>
+            /// <param name="isEmbedded">A flag indicating whether the object is embedded or not.</param>
+            /// <seealso cref="EmbeddedObject"/>
+            /// <exception cref="ArgumentNullException">Thrown if <paramref name="name"/> is <c>null</c>.</exception>
+            /// <exception cref="ArgumentException">Thrown if <paramref name="name"/> is the empty string.</exception>
+            [Obsolete("Use Builder(string name, ObjectSchemaType schemaType) instead.")]
+            public Builder(string name, bool isEmbedded = false) : this(
+                name,
+                isEmbedded ? ObjectSchemaType.Embedded : ObjectSchemaType.TopLevel)
```
Comment (nirinchev) [inline Realm/Realm/Schema/ObjectSchema.cs:204]:
```suggestion
            public Builder(string name, bool isEmbedded = false)
                : this(name, isEmbedded ? ObjectSchemaType.Embedded : ObjectSchemaType.TopLevel)
```

File: Realm/Realm/Schema/ObjectSchema.cs [hunk 13]
```diff
@@ -145,31 +165,54 @@ public class Builder : SchemaBuilderBase<Property>
...
+            // TODO andrea: review the documentation
             /// <summary>
             /// Initializes a new instance of the <see cref="Builder"/> class populated with properties from the
             /// provided <paramref name="type"/>.
             /// </summary>
             /// <param name="type">
-            /// The <see cref="System.Type"/> that will be used to populate the builder. It must be a <see cref="RealmObject"/>
-            /// or <see cref="EmbeddedObject"/> inheritor.
+            /// The <see cref="System.Type"/> that will be used to populate the builder. It must be a <see cref="RealmObject"/>,
+            /// an <see cref="EmbeddedObject"/>  or a <see cref="AsymmetricObject"/>inheritor.
```
Comment (nirinchev) [inline Realm/Realm/Schema/ObjectSchema.cs:215]:
```suggestion
            /// an <see cref="EmbeddedObject"/> , or a <see cref="AsymmetricObject"/>inheritor.
```

File: Realm/Realm/Schema/ObjectSchema.cs [hunk 14]
```diff
@@ -212,10 +255,13 @@ public Builder(string name, bool isEmbedded = false)
             public Builder(Type type)
             {
                 Argument.NotNull(type, nameof(type));
-                Argument.Ensure(type.IsRealmObject() || type.IsEmbeddedObject(), $"The class {type.FullName} must descend directly from RealmObject or EmbeddedObject", nameof(type));
+                Argument.Ensure(type.IsRealmObject() || type.IsEmbeddedObject() || type.IsAsymmetricObject(),
+                    $"The class {type.FullName} must descend directly from either RealmObject, EmbeddedObject or AsymmetricObject", nameof(type));
```
Comment (nirinchev) [inline Realm/Realm/Schema/ObjectSchema.cs:259]:
```suggestion
                    $"The class {type.FullName} must descend directly from either RealmObject, EmbeddedObject, or AsymmetricObject", nameof(type));
```

File: Realm/Realm/Schema/ObjectSchema.cs [hunk 15]
```diff
@@ -145,31 +165,54 @@ public class Builder : SchemaBuilderBase<Property>
             /// Gets or sets a value indicating whether this <see cref="Builder"/> describes an embedded object.
             /// </summary>
             /// <value><c>true</c> if the schema pertains to an <see cref="EmbeddedObject"/> instance; <c>false</c> otherwise.</value>
+            [Obsolete("Check against RealmSchemaType instead.")]
             public bool IsEmbedded { get; set; }
```
Comment (nirinchev) [inline Realm/Realm/Schema/ObjectSchema.cs:169]: This should be implemented in terms of `RealmSchemaType`.

**Review by nirinchev — COMMENTED** [review]
Haven't reviewed the tests yet, but so far looks pretty good. I've mostly have nitty comments.

File: Realm/Realm/Schema/ObjectSchema.cs (hunk 11 above)
Comment (LaPeste) [reply Realm/Realm/Schema/ObjectSchema.cs:41]: Agree. How about one of the following, in order of my personal preference?
1. `RealmObjectType`
1. `BaseObjectType`
1. `ObjectType`

**Comment by LaPeste** [comment]
@nirinchev you can start reviewing the PR. The only thing missing is that I want to see if I can include more test values in `AsymmetricObjectTests.SetAndRemotelyReadValue`. This test is already working, beside the min and max `Decimal128`, which is should be related to the [bug](https://github.com/realm/realm-core/pull/5740) in core you mentioned.

File: Realm/Realm/Realm.cs [hunk 16]
```diff
@@ -595,7 +595,81 @@ public void Add<T>(IEnumerable<T> objs, bool update = false)
...
+        /// <exception cref="ObjectDisposedException">
+        /// If you invoke this method on a closed <see cref="Realm"/>.
+        /// </exception>
+        /// <exception cref="ArgumentNullException">
+        /// If you pass a null argument to this method.
+        /// </exception>
+        /// <remarks>
+        /// If the object is already managed by this <see cref="Realm"/>, this method does nothing.
+        /// This method modifies the object in-place, meaning that after it has run, <see cref="AsymmetricObject"/> will be managed.
+        /// Once an <see cref="AsymmetricObject"/> becomes managed dereferencing the original <see cref="AsymmetricObject"/>
```
Comment (nirinchev) [inline Realm/Realm/Realm.cs:625]: I don't understand this part.

File: Realm/Realm/Realm.cs [hunk 17]
```diff
@@ -608,8 +682,14 @@ internal void ManageEmbedded(IEmbeddedObject obj, ObjectHandle handle)
             obj.SetManagedAccessor(new ManagedAccessor(this, handle, metadata), metadata.Helper, false, true);
         }
 
-        private void AddInternal(IRealmObject obj, Type objectType, bool update)
+        private void AddInternal<T>(T obj, Type objectType, bool update)
+            where T : IRealmObjectBase
         {
+            if (obj.GetType().IsEmbeddedObject())
```
Comment (nirinchev) [inline Realm/Realm/Realm.cs:688]:
```suggestion
            if (objectType.IsEmbeddedObject())
```

File: Realm/Realm/Schema/ObjectSchema.cs (hunk 11 above)
Comment (nirinchev) [reply Realm/Realm/Schema/ObjectSchema.cs:41]: Let's go with `ObjectType`.

File: Tests/Realm.Tests/Database/DynamicAsymmetricObjectTests.cs [hunk 18]
```diff
@@ -0,0 +1,87 @@
+﻿// ////////////////////////////////////////////////////////////////////////////
```
Comment (nirinchev) [inline Tests/Realm.Tests/Database/DynamicAsymmetricObjectTests.cs:1]: Double comments

File: Tests/Realm.Tests/Database/DynamicAsymmetricObjectTests.cs [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+using Realms.Tests.Sync;
+
+namespace Realms.Tests.Database
+{
+    [TestFixture, Preserve(AllMembers = true)]
+    public class DynamicAsymmetricObjectTests : SyncTestBase
+    {
+        [TestCase(true)]
+        [TestCase(false)]
+        public void ReadWriteVarsInWriteTransaction(bool isDynamic)
```
Comment (nirinchev) [inline Tests/Realm.Tests/Database/DynamicAsymmetricObjectTests.cs:31]:
```suggestion
        [Test]
        public void ReadWriteVarsInWriteTransaction([Values(true, false)] bool isDynamic)
```

File: Tests/Realm.Tests/Sync/SyncTestBase.cs [hunk 20]
```diff
@@ -197,10 +197,10 @@ protected async Task<FlexibleSyncConfiguration> GetFLXIntegrationConfigAsync(App
             return UpdateConfig(new FlexibleSyncConfiguration(user, optionalPath));
         }
 
-        protected async Task<Realm> GetFLXIntegrationRealmAsync(App app = null)
+        protected async Task<Realm> GetFLXIntegrationRealmAsync(App app = null, FlexibleSyncConfiguration flxConfig = null)
         {
-            var config = await GetFLXIntegrationConfigAsync(app);
-            return await GetRealmAsync(config);
+            flxConfig ??= await GetFLXIntegrationConfigAsync(app);
+            return await GetRealmAsync(flxConfig);
```
Comment (nirinchev) [inline Tests/Realm.Tests/Sync/SyncTestBase.cs:203]: Why this change? If you pass in the config, it's better to just use `GetRealmAsync` directly.

File: Tests/Realm.Tests/Database/DynamicAsymmetricObjectTests.cs [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+        {
+            SyncTestHelpers.RunBaasTestAsync(async () =>
+            {
+                var flxConfig = await GetFLXIntegrationConfigAsync();
+                flxConfig.IsDynamic = isDynamic;
+                flxConfig.Schema = new[] { typeof(AsymmetricObjectWithAllTypes) };
+                using var realm = await GetFLXIntegrationRealmAsync(flxConfig: flxConfig);
```
Comment (nirinchev) [inline Tests/Realm.Tests/Database/DynamicAsymmetricObjectTests.cs:38]:
```suggestion
                using var realm = await GetRealmAsync(flxConfig);
```

File: Tests/Realm.Tests/Database/ObjectSchemaTests.cs [hunk 22]
```diff
@@ -884,7 +891,8 @@ public void ObjectSchema_GetBuilder_ContainsAllProperties()
 
             Assert.That(originalSchema, Is.EquivalentTo(builder));
             Assert.That(originalSchema.Name, Is.EqualTo(builder.Name));
-            Assert.That(originalSchema.IsEmbedded, Is.EqualTo(builder.IsEmbedded));
+            Assert.That(originalSchema.ObjectType == ObjectSchema.ObjectSchemaType.EmbeddedObject,
+                Is.EqualTo(builder.RealmSchemaType == ObjectSchema.ObjectSchemaType.EmbeddedObject));
```
Comment (nirinchev) [inline Tests/Realm.Tests/Database/ObjectSchemaTests.cs:895]: This seems like a weird check - why not just assert that `originalSchema.ObjectType, Is.EqualTo(builder.RealmSchemaType)`?

File: Tests/Weaver/Realm.Fody.Tests/WeaverTests.cs [hunk 23]
```diff
@@ -521,7 +522,16 @@ public void MatchErrorsAndWarnings()
                 "RealmSetWithSetter.People has a setter but its type is a ISet which only supports getters.",
                 "RealmDictionaryWithSetter.People has a setter but its type is a IDictionary which only supports getters.",
                 "RealmDictionaryWithNonStringKey.People is a Dictionary<Int32, Person> but only string keys are currently supported by Realm.",
-                "MixOfCollectionsObject.EmbeddedSet is a Set<EmbeddedObject> which is not supported. Embedded objects are always unique which is why List<EmbeddedObject> already has Set semantics."
+                "MixOfCollectionsObject.EmbeddedSet is a Set<EmbeddedObject> which is not supported. Embedded objects are always unique which is why List<EmbeddedObject> already has Set semantics.",
+                "Measurement.Sensor is of type AsymmetricObject, but AsymmetricObject-s aren't allowed to be the receiving end of any relationship.",
```
Comment (nirinchev) [inline Tests/Weaver/Realm.Fody.Tests/WeaverTests.cs:526]: Why is there a dash here?

File: Tools/DeployApps/BaasClient.cs [hunk 24]
```diff
@@ -699,6 +717,17 @@ private static class Schemas
...
+            private static object GenericFlxBaasRule(string differentiator, string collectionName) => new
+            {
+                collection = collectionName,
+                database = $"FLX_{differentiator}",
+                roles = new[] { _flxDefaultRoles }
+            };
+
+            public static object AsymmetricObjectWithAllTypesBaasRule(string differentiator) => GenericFlxBaasRule(differentiator, "AsymmetricObjectWithAllTypes");
+
+            public static object HugeSyncAsymmetricObjectBaasRule(string differentiator) => GenericFlxBaasRule(differentiator, "HugeSyncAsymmetricObject");
```
Comment (nirinchev) [inline Tools/DeployApps/BaasClient.cs:729]: We probably don't need these explicit functions, we can use the common one directly. Also, we should use `nameof` here instead of hardcoded strings.

File: Tests/Realm.Tests/Database/ObjectSchemaTests.cs [hunk 25]
```diff
@@ -25,6 +25,8 @@
 
 namespace Realms.Tests.Database
 {
+#pragma warning disable CS0618 // ObjectSchema.Builder(string, bool) is obsolete, Builder(string, ObjectSchemaType) should be used instead
```
Comment (nirinchev) [inline Tests/Realm.Tests/Database/ObjectSchemaTests.cs:28]: Do we need to disable this for the entire file? Can we migrate all the tests to use the new constructor and if you want to test the old one, then create a new test and mark it `[Obsolete]`?

File: Tests/Realm.Tests/Database/AsymmetricObjectTests.cs [hunk 26]
```diff
@@ -0,0 +1,382 @@
...
+                    realm.Write(() =>
+                    {
+                        realm.Add(asymmetricObj);
+                    });
+
+                    realm.Write(() =>
+                    {
+                        realm.Add(asymmetricObj);
+                    });
+                });
```
Comment (nirinchev) [inline Tests/Realm.Tests/Database/AsymmetricObjectTests.cs:216]: Why doesn't this throw? After the first write, we detach the object, then try to add it again - this should be an error as we're trying to add a deleted object to the Realm.

File: Tests/Realm.Tests/Database/AsymmetricObjectTests.cs [hunk 27]
```diff
@@ -0,0 +1,382 @@
...
+            });
+        }
+
+        [TestCaseSource(nameof(SetAndGetValueCases))]
+        [TestCaseSource(nameof(SetAndReplaceWithNullCases))]
+        public void SetAndRemotelyReadValue(string propertyName, object propertyValue)
+        {
+            SyncTestHelpers.RunBaasTestAsync(async () =>
+            {
+                Guid objId = Guid.Empty;
```
Comment (nirinchev) [inline Tests/Realm.Tests/Database/AsymmetricObjectTests.cs:233]:
```suggestion
                var objId = Guid.Empty;
```

File: Tests/Realm.Tests/Database/AsymmetricObjectTests.cs [hunk 28]
```diff
@@ -0,0 +1,382 @@
...
+                });
+
+                await realm.SyncSession.WaitForUploadAsync();
+                var documents = await GetObjFromRemoteThroughMongoClient<AsymmetricObjectWithAllTypes>(
+                    flxConfig, propertyName, BsonValue.Create(propertyValue));
+
+                foreach (var doc in documents)
+                {
+                    Assert.That(TestHelpers.GetPropertyValue(doc, propertyName), Is.EqualTo(propertyValue));
+                }
```
Comment (nirinchev) [inline Tests/Realm.Tests/Database/AsymmetricObjectTests.cs:254]: This assertion is weird. Since you're querying for objects with that property value, it's normal that all the returned documents will have the correct property. Instead, you should probably query by the objId and then assert on the property value.

File: Tests/Realm.Tests/Database/AsymmetricObjectTests.cs [hunk 29]
```diff
@@ -0,0 +1,382 @@
...
+                        realm.Add(new PrimaryKeyInt32Object
+                        {
+                            Id = id
+                        });
+                    });
+                });
+            });
+        }
+
+        [Test, NUnit.Framework.Explicit("Once Daniel Tabacaru's work is done on \"error actions\" this will be an \"application bug\" action: https://github.com/10gen/baas/blob/9f32d54aa79aff6dfb36a6c07742594e38b07441/realm/sync/protocol/protocol_errors.go#L439")]
```
Comment (nirinchev) [inline Tests/Realm.Tests/Database/AsymmetricObjectTests.cs:294]: This should be in `main` now.

File: Tests/Realm.Tests/Database/AsymmetricObjectTests.cs [hunk 30]
```diff
@@ -0,0 +1,382 @@
...
+        {
+            TestHelpers.RunAsyncTest(async () =>
+            {
+                var config = (RealmConfiguration)RealmConfiguration.DefaultConfiguration;
+                config.Schema = new[] { typeof(BasicAsymmetricObject) };
+
+                Exception ex = null;
+                try
+                {
+                    using var realm = await GetRealmAsync(config);
```
Comment (nirinchev) [inline Tests/Realm.Tests/Database/AsymmetricObjectTests.cs:327]: Why do you need the async version here - wouldn't you get the same error with `GetRealm`?

File: Tests/Realm.Tests/Database/AsymmetricObjectTests.cs [hunk 31]
```diff
@@ -0,0 +1,382 @@
...
+                {
+                    ex = e;
+                }
+
+                Assert.That(ex, Is.InstanceOf<RealmSchemaValidationException>());
+                Assert.That(ex.Message.Contains($"Asymmetric table \'{nameof(BasicAsymmetricObject)}\'"), Is.True);
+            });
+        }
+
+        private static Task<T[]> GetObjFromRemoteThroughMongoClient<T>(FlexibleSyncConfiguration config, string remoteFieldName, BsonValue fieldValue, string mongoClientCondition = MongoClientCondition.Equality)
```
Comment (nirinchev) [inline Tests/Realm.Tests/Database/AsymmetricObjectTests.cs:339]:
```suggestion
        private static Task<T[]> GetObjFromRemoteThroughMongoClient<T>(User user, string remoteFieldName, BsonValue fieldValue, string mongoClientCondition = MongoClientCondition.Equality)
```

I thought it's quite confusing we're passing a config here.

File: Realm/Realm/Realm.cs (hunk 16 above)
Comment (LaPeste) [reply Realm/Realm/Realm.cs:625]: Could you be more specific?

File: Realm/Realm/Schema/ObjectSchema.cs (hunk 11 above)
Comment (LaPeste) [reply Realm/Realm/Schema/ObjectSchema.cs:41]: Then how would you like the field called? It's currently called exactly `ObjectType`. We could call it `SchemaType`? Would that be ok?

File: Realm/Realm/Realm.cs (hunk 16 above)
Comment (nirinchev) [reply Realm/Realm/Realm.cs:625]: "dereferencing the original reference throws an exception" - I don't believe this is the case, why would it throw?

Comment (LaPeste) [reply Realm/Realm/Realm.cs:625]: Once the `AsymmetricObject` becomes managed, you get a runtime error if trying to use that reference.
We all agreed that this is how it should be. Why do you disagree?
The test for this is `AccessAsymmetricObjAfterAddedToRealm_Throws`

File: Tests/Realm.Tests/Sync/SyncTestBase.cs (hunk 20 above)
Comment (LaPeste) [reply Tests/Realm.Tests/Sync/SyncTestBase.cs:203]: I don't understand this comment. If the config is passed, it already directly uses `GetRealmAsync`.

File: Tests/Weaver/Realm.Fody.Tests/WeaverTests.cs (hunk 23 above)
Comment (LaPeste) [reply Tests/Weaver/Realm.Fody.Tests/WeaverTests.cs:526]: We usually write with the back-ticks, where it's easy to understand how the object is called -> `AsymmetricObject`s.
 I wanted to keep it simple to read when not having the back-ticks in place. Nothing else.
Should I just have AsymmetricObjects?

File: Realm/Realm/Realm.cs (hunk 16 above)
Comment (LaPeste) [reply Realm/Realm/Realm.cs:625]: Conversation taken offline. Resolved.

File: Tests/Realm.Tests/Database/AsymmetricObjectTests.cs (hunk 30 above)
Comment (LaPeste) [reply Tests/Realm.Tests/Database/AsymmetricObjectTests.cs:336]: Wouldn't this be true for almost any test? Is there a technical reason you're asking this question on this specific test?
And in general, unless one is testing the async API, what would be the reason to want to specifically use one (the async one) over the other (the sync one)?
(I assume you're asking here because there's no need to use the async method to grab the configuration)

File: CHANGELOG.md [hunk 32]
```diff
@@ -1,6 +1,46 @@
 ## vNext (TBD)
 
 ### Enhancements
+* Introduced `AsymmetricObject` intended for write-heavy workloads, where high performance is generally important. This new object:
+  1. syncs data unidirectionaly, from the clients to the server
+  1. can't be queried, deleted or modified once added to the Realm
+  1. and as a result the previous point, at the end of a write transaction dereferencing the added `AsymmetricObject` throws an exception
+  1. is only usable with flexible sync
+  1. can't be the receiveing end of any type of relationship
+  1. can only have a one-to-many relationship with `EmbeddedObject` and collections of the latter
```
Comment (nirinchev) [inline CHANGELOG.md:10]: It can also a have -to-one relationships.

File: CHANGELOG.md [hunk 33]
```diff
@@ -1,6 +1,46 @@
...
+
+  In the same write transaction, it is perfetcly legal to add `AsymmetricObject`-s and `RealmObject`-s
```
Comment (nirinchev) [inline CHANGELOG.md:12]:
```suggestion
  In the same write transaction, it is perfetcly legal to add `AsymmetricObject`s and `RealmObject`s
```

File: CHANGELOG.md [hunk 34]
```diff
@@ -1,6 +1,46 @@
...
+
+  //.....
+
+  var asymmetricObject = new BasicAsymmetricObject();
+
+  realm.Write(() =>
+  {
+      realm.Add(asymmetricObject);
+
+      realm.Add(new PrimaryKeyInt32Object
```
Comment (nirinchev) [inline CHANGELOG.md:35]: Should this be `Person`?

File: Realm/Realm.Weaver/RealmWeaver.cs [hunk 35]
```diff
@@ -381,13 +381,21 @@ private WeavePropertyResult WeaveProperty(PropertyDefinition prop, TypeDefinitio
...
 
             var backlinkAttribute = prop.CustomAttributes.FirstOrDefault(a => a.AttributeType.Name == "BacklinkAttribute");
-            if (backlinkAttribute != null && !prop.IsIQueryable())
+            if (backlinkAttribute != null)
             {
-                return WeavePropertyResult.Error($"{type.Name}.{prop.Name} has [Backlink] applied, but is not IQueryable.");
+                if (!prop.IsIQueryable())
+                {
+                    return WeavePropertyResult.Error($"{type.Name}.{prop.Name} has [Backlink] applied, but is not IQueryable.");
+                }
```
Comment (nirinchev) [inline Realm/Realm.Weaver/RealmWeaver.cs:398]: Why was that split?

File: Realm/Realm/Realm.cs [hunk 36]
```diff
@@ -595,7 +595,85 @@ public void Add<T>(IEnumerable<T> objs, bool update = false)
...
+        /// dereferencing any property of the original <see cref="AsymmetricObject"/> reference throw an exeception.
+        /// Hence, none of the properties of the elements in the collection can be deferenced anymore after the transaction.
+        /// </remarks>
+        public void Add<T>(IEnumerable<T> objs)
+            where T : IAsymmetricObject
+        {
+            ThrowIfDisposed();
+            Argument.NotNull(objs, nameof(objs));
+            Argument.Ensure(objs.All(o => o != null), $"{nameof(objs)} must not contain null values.", nameof(objs));
+            Argument.Ensure(objs.All(o => !o.IsManaged), $"{nameof(objs)} must not contain already managed objects by a Realm.", nameof(objs));
```
Comment (nirinchev) [inline Realm/Realm/Realm.cs:672]: This is inefficient because we're iterating the collection twice. We could instead have a foreach loop that checks for both conditions.

File: Realm/Realm/Schema/ObjectSchema.cs [hunk 37]
```diff
@@ -145,31 +174,64 @@ public class Builder : SchemaBuilderBase<Property>
...
+            {
+                get
+                {
+                    return RealmSchemaType == ObjectType.EmbeddedObject;
+                }
+
+                set
+                {
+                    // being an obsolete method, it's assumed that it's old code that doesn't use AsymmetricObjects
+                    _ = value ? RealmSchemaType = ObjectType.EmbeddedObject : RealmSchemaType = ObjectType.RealmObject;
```
Comment (nirinchev) [inline Realm/Realm/Schema/ObjectSchema.cs:188]: That is super awkward - let's simplify to

```suggestion
                    RealmSchemaType = value ? ObjectType.EmbeddedObject : ObjectType.RealmObject;
```

**Review by nirinchev — COMMENTED** [review]
Final nits from me

File: Realm/Realm.Weaver/RealmWeaver.cs (hunk 35 above)
Comment (LaPeste) [reply Realm/Realm.Weaver/RealmWeaver.cs:398]: A forgotten leftover of a check that later was moved down. Thank you for noticing it.

File: Realm/Realm/Schema/ObjectSchema.cs (hunk 37 above)
Comment (LaPeste) [reply Realm/Realm/Schema/ObjectSchema.cs:188]: Emh.... 🤦🏻
